### PR TITLE
Use raw pointers for std::sort if possible

### LINKED ISSUE #7263: Kokkos::sort using std::sort is 2x slower than std::sort
With
```C++
#include <Kokkos_Core.hpp>
#include <Kokkos_Random.hpp>
#include <Kokkos_Sort.hpp>

#include <cstdio>
#include <iostream>

int main(int argc, char* argv[]) {
  Kokkos::ScopeGuard scope_guard;

  for (int n=1; n<1'000'000'000; n*=10) {
    Kokkos::View<float*> bla("bla10", n);
    Kokkos::Random_XorShift64_Pool<> g(1931);
    Kokkos::fill_random(bla, g, 100.0);

    Kokkos::View<float*> blabla("bla10", n);
    Kokkos::deep_copy(blabla, bla);

    Kokkos::Timer timer;
    {
      double start = timer.seconds();
      Kokkos::sort(bla);
     double stop = timer.seconds();
      std::cout << n << " Kokkos::sort: " << stop-start << "s\n";
    }
    {
      double start = timer.seconds();
      std::sort(blabla.data(), blabla.data()+blabla.size());
      double stop = timer.seconds();
      std::cout << n << " std::sort: " << stop-start << "s\n";
    }
  }
}
```
I'm seeing
```
1 Kokkos::sort: 1.25e-07s
1 std::sort: 2.91e-07s
10 Kokkos::sort: 2.5458e-05s
10 std::sort: 3.75e-07s
100 Kokkos::sort: 1.5333e-05s
100 std::sort: 4.917e-06s
1000 Kokkos::sort: 0.000176583s
1000 std::sort: 6.2667e-05s
10000 Kokkos::sort: 0.00208267s
10000 std::sort: 0.000840209s
100000 Kokkos::sort: 0.0207976s
100000 std::sort: 0.00741763s
1000000 Kokkos::sort: 0.15895s
1000000 std::sort: 0.0668978s
10000000 Kokkos::sort: 1.78158s
10000000 std::sort: 0.79071s
100000000 Kokkos::sort: 19.5198s
100000000 std::sort: 9.00485s
```
It seems going through random access iterators is much slower than giving the raw memory to `std::sort`.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/7263.

### CODE REVIEW

File: algorithms/src/sorting/impl/Kokkos_SortImpl.hpp [hunk 1]
```diff
@@ -268,19 +268,29 @@ void copy_to_host_run_stdsort_copy_back(
...
     // run sort on the mirror of view_dc
-    auto mv_h  = create_mirror_view_and_copy(Kokkos::HostSpace(), view_dc);
-    auto first = KE::begin(mv_h);
-    auto last  = KE::end(mv_h);
-    std::sort(first, last, std::forward<MaybeComparator>(maybeComparator)...);
+    auto mv_h = create_mirror_view_and_copy(Kokkos::HostSpace(), view_dc);
+    if (view.span_is_contiguous()) {
+      std::sort(mv_h.data(), mv_h.data() + mv_h.size(),
+                std::forward<MaybeComparator>(maybeComparator)...);
+    } else {
```
Comment (dalg24) [inline algorithms/src/sorting/impl/Kokkos_SortImpl.hpp:275]: Please comment (here in reply) what this added complexity brings to the test?

Comment (masterleinad) [reply algorithms/src/sorting/impl/Kokkos_SortImpl.hpp:275]: What test do you mean?

Comment (dalg24) [reply algorithms/src/sorting/impl/Kokkos_SortImpl.hpp:275]: Did not read it right. Sorry.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
Would you share your updated performance numbers ?

**Comment by masterleinad** [comment]
> Would you share your updated performance numbers ?

See https://github.com/kokkos/kokkos/issues/7263#issuecomment-2310868537.

**Review by Rombur — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring unrelated failures (GPU not detected at runtime in Clang+CUDA build and host in host device warning treated as error)

**Pull request merged**